Enables arbitrary execute flags

aeslaughter opened the pull request:
(closes #7489)
<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

aeslaughter committed:
Enables arbitrary execute flags

(closes #7489)

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/125740/) on 0a250bd wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 54d9a66c54b8c2f820e5eab899bb9aebf5dd076f`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10110/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/10110/style.patch | git apply -v`

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/125822/) on 4000506 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/10110/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](<WSGIRequest: POST '/client/job_finished/1051070556/buildq7_0/127509/'>) on 8b35f24 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 6067c3c36590fbe2bce980fe989256732cb38e79`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10110/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/10110/style.patch | git apply -v`

aeslaughter commented:
@permcody This is ready for review, I will work on fixing up the apps after the design here is final. I don't want to spend time changing apps if the API is going to change.

permcody reviewed (dismissed):
For the most part this looks OK. I have a few issues with the design though. I'm not sold on the global constants that you are defining to create new enum items, but I don't have a better suggestion at the moment. I know you are trying to get away from strings as extra enums. Anyway, let's chat this over and see if we can make a decision.

aeslaughter commented on framework/include/utils/MooseEnumBase.h:
```diff
@@ -109,11 +95,39 @@ class MooseEnumBase
+  void addEnumerationNames(const std::string & names);
+  void addEnumerationName(const std::string & raw_name);
+  void addEnumerationName(const std::string & name, const int & value);
+  void addEnumerationName(const MooseEnumItem & item);
```
These should be protected in the base and are basically just a replacement for the filenames method that was there, using several methods so that extra flexibility is available down the line.

aeslaughter commented on framework/include/base/AuxGroupExecuteMooseObjectWarehouse.h:
```diff
@@ -93,9 +93,9 @@ AuxGroupExecuteMooseObjectWarehouse<T>::updateDependObjects(
-  constexpr std::uint16_t all_flags = std::numeric_limits<std::uint16_t>::max();
+  const std::uint16_t initial_flag_mask = static_cast<std::uint16_t>(EXEC_INITIAL);
+  const std::uint16_t not_initial_flag_mask = ~static_cast<std::uint16_t>(EXEC_INITIAL);
+  const std::uint16_t all_flags = std::numeric_limits<std::uint16_t>::max();
```
It wouldn't compile, using the new MooseEnumItem I think might get in the way of the  restriction that the "type must be a LiteralType."

aeslaughter replied:
```
../framework/include/base/AuxGroupExecuteMooseObjectWarehouse.h:97:79: note: non-constexpr function 'operator int' cannot be used in a constant expression
```

aeslaughter commented on framework/include/utils/MooseUtils.h:
```diff
@@ -55,6 +60,11 @@ void escape(std::string & str);
 /**
+ * Python like join method.
+ */
+std::string join(const std::vector<std::string> & values, const std::string & sep = ", ");
```
Correct, I forgot to update this.

aeslaughter commented on framework/include/utils/MultiMooseEnum.h:
```diff
@@ -16,17 +16,27 @@
 
+namespace MooseUtils
+{
+MultiMooseEnum createExecuteOnEnum(const std::set<ExecFlagType> &,
```
I will add it. There should be some discussion in the markdown. This replaces the SetupInterface::getExecuteFlags method.

aeslaughter replied:
Nevermind, this is left-over from another version. I will pull this out.

aeslaughter commented on framework/src/executioners/EigenExecutionerBase.C:
```diff
@@ -428,14 +428,13 @@ EigenExecutionerBase::normalizeSolution(bool force)
+    _problem.execute(EXEC_NONLINEAR);
+    _problem.execute(EXEC_LINEAR);
+    _problem.execute(EXEC_TIMESTEP_END);
+    _problem.execute(EXEC_SUBDOMAIN);
```
This could probably loop over the global Moose::execute_enum, this was done in an early version that didn't have an global list of available flags.

aeslaughter commented on framework/src/controls/Control.C:
```diff
@@ -23,9 +23,11 @@ validParams<Control>()
   params += validParams<FunctionInterface>();
+
+  params.set<ExecFlagEnum>("execute_on", true) =
+      std::vector<ExecFlagType>({EXEC_INITIAL, EXEC_TIMESTEP_END});
```
It is just a MultiMooseEnum, people have been using MultiMooseEnum for execute flags for some time.

You can still set them as: `params.set<ExecFlagEnum>("execute_on") = "foo bar";` if you prefer, I was just trying to avoid using strings.

aeslaughter commented on framework/src/outputs/AdvancedOutput.C:
```diff
@@ -538,19 +541,22 @@ AdvancedOutput::initOutputList(OutputData & data)
 AdvancedOutput::addValidParams(InputParameters & params, const MultiMooseEnum & types)
 {
+  ExecFlagEnum empty_execute_on = MooseUtils::getDefaultExecFlagEnum();
+  empty_execute_on.addAvailableFlags({EXEC_FINAL, EXEC_FAILED});
```
I add a version like that, but didn't like it either. From a user perspective it is even more confusing. But, I can try it.

aeslaughter commented on framework/src/utils/MooseEnumBase.C:
```diff
@@ -54,58 +48,97 @@ MooseEnumBase::MooseEnumBase(const MooseEnumBase & other_enum)
+  // strtol allows for proper conversions of both int and hex strings
+  int value;
+  if (name_value.size() == 2)
+    value = strtol(name_value[1].c_str(), NULL, 0);
```
OK

aeslaughter commented on modules/contact/src/ContactPenetrationAuxAction.C:
```diff
@@ -45,10 +45,7 @@ ContactPenetrationAuxAction::act()
-    MultiMooseEnum execute_options = SetupInterface::getExecuteOptions();
-    execute_options = "initial linear";
-    params.set<MultiMooseEnum>("execute_on") = execute_options;
+    params.set<ExecFlagEnum>("execute_on") = std::vector<ExecFlagType>({EXEC_INITIAL, EXEC_LINEAR});
```
In general, I maintained the use of quite mode. For specific objects like this it probably is designed to only work with these flags (it should probably be suppressed after setting). But, for system level changes (i.e., Output) quiet mode should be used because we are just setting sensible defaults not trying to dictate proper behavior.

aeslaughter commented on modules/contact/src/ContactPressureAuxAction.C:
```diff
@@ -65,9 +65,8 @@ ContactPressureAuxAction::act()
-    execute_options = "nonlinear timestep_end timestep_begin";
-    params.set<MultiMooseEnum>("execute_on") = execute_options;
+    params.set<ExecFlagEnum>("execute_on", true) =
+        std::vector<ExecFlagType>({EXEC_NONLINEAR, EXEC_TIMESTEP_END, EXEC_TIMESTEP_BEGIN});
```
That was my original idea, but it is ambiguous with other operator= calls. However, I could probably only define this version in ExecFlagEnum of the operator= and that might be possible. 

I will have to play with that a bit.

aeslaughter commented on modules/level_set/src/base/LevelSetTypes.C:
```diff
@@ -0,0 +1,14 @@
+// MOOSE includes
+#include "MooseEnumItem.h"
+
+const ExecFlagType LevelSet::EXEC_ADAPT_MESH("ADAPT_MESH", 8000);
```
Perhaps.

permcody replied:
Would still like to see this addressed. It's one of the weird unfinished things since it requires the user to produce something and they may not know how to do that.

aeslaughter commented on test/src/postprocessors/TestPostprocessor.C:
```diff
@@ -56,6 +61,11 @@ TestPostprocessor::getValue()
+  else if (_test_type == "custom_execute_on")
+  {
+    _console << "Flag Name: " << EXEC_JUST_GO << std::endl;
+    return _execute_count++;
```
ok

aeslaughter commented:
@bwspenc Let me know when you think you are ready for this or if you need some help.

aeslaughter commented on framework/src/utils/MultiMooseEnum.C:
```diff
@@ -124,6 +124,31 @@ MultiMooseEnum::operator=(const std::set<std::string> & names)
 }
 
+MultiMooseEnum &
+MultiMooseEnum::operator=(const std::vector<MooseEnumItem> & items)
```
No. This just sets the supplied items to the current list. The same as doing `me = "foo bar"`.

aeslaughter committed:
Re-factor assignment operator for ExecFlagEnum to use initializer_list

(refs #7489)

aeslaughter committed:
Re-factor add/remove methods in ExecFlagEnum to use variadic arguments

(refs #7489)

aeslaughter committed:
Add += operator to ExecFlagEnum

(refs #7489)

aeslaughter committed:
Address review comments

(refs #7489)

aeslaughter committed:
Allow for ExecFlagType to be created without assigning id

(refs #7489)

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/129349/) on e0ca19d wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format b03ec4b32400d092a5239e91b975fcfe4a78675b`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10110/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/10110/style.patch | git apply -v`

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/129360/) on db93b5e wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format b03ec4b32400d092a5239e91b975fcfe4a78675b`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10110/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/10110/style.patch | git apply -v`

aeslaughter committed:
Simplify the default enum generator

(refs #7489)

aeslaughter committed:
Allows applications to continue to use MultiMooseEnum for "execute_on"

Most of these changes will be removed or deprecated when applications
are updated to use the new ExecFlagEnum type.

(refs #7489)

aeslaughter committed:
Update to make EXEC_FINAL a default execute flag

(refs #7489)

aeslaughter commented:
@permcody This is ready for another review; however, we should probably wait to merge this until after 🦃.

permcody commented on framework/src/outputs/AdvancedOutputUtils.C:
```diff
@@ -18,31 +18,34 @@
+  if (parameters.have_parameter<ExecFlagEnum>("execute_system_information_on"))
+  {
+    _map.insert(std::make_pair("system_information", MooseUtils::getDefaultExecFlagEnum()));
+    _map["system_information"] = EXEC_INITIAL;
```
Wait, two inserts for the same key? Maybe you have an overload so that the option just gets changed but it looks screwy. I'd prefer you build the right value and insert once.

aeslaughter replied:
Not two inserts, the _map.insert creates the entry. The second line sets the current flag to EXEC_INITIAL, using the operator= from the MultiMooseEnum class.

permcody replied:
OK - so you are using the overloaded operator here. I think it would be much clearer to just build the value you need, then insert it.

aeslaughter replied:
I can do that.

permcody commented on framework/src/outputs/Output.C:
```diff
@@ -82,12 +79,15 @@ validParams<Output>()
+  ::mooseDeprecated("The 'getExecuteOptions' was replaced by the ExecFlagEnum class because MOOSE "
+                    "was updated to use this for the execute flags and the new function provides "
+                    "additional arguments for modification of the enum.");
+  */
```
Remove?

aeslaughter replied:
These will be enabled after the apps have been updated, so non tested apps will get a useful message. I can pull the comments out and add them back in a non-comments when the update is finished, it was just easier to put them in while I was working on this and then comment them out to get everything working.

permcody commented on framework/src/utils/ExecFlagEnum.C:
```diff
@@ -0,0 +1,96 @@
+                  "available options include \'");
+  for (const std::string & name : getNames())
+    doc += name + "', '";
+  doc.erase(doc.end() - 4, doc.end());
```
Umm....   Try using the new container stringify method. It should clean this whole mess up.

permcody commented on framework/src/utils/ExecFlagEnum.C:
```diff
@@ -0,0 +1,96 @@
+ExecFlagEnum::operator=(const std::initializer_list<ExecFlagType> & flags)
+{
+  clear();
+  (*this) += flags;
```
Minor: Parens are unecessary

permcody commented on test/src/base/MooseTestApp.C:
```diff
@@ -641,3 +642,10 @@ MooseTestApp::associateSyntax(Syntax & syntax, ActionFactory & action_factory)
+MooseTestApp::registerExecFlags()
+{
+  MooseApp::registerExecFlags();
+  registerExecFlag(EXEC_JUST_GO);
```
Oh so you did take care of the number problem!

permcody reviewed (dismissed):
Getting closer. Go through this again and pay attention to the extra headers you've included in several places and lets see if we can drop new friend classes.

aeslaughter commented on framework/include/base/SetupInterface.h:
```diff
@@ -62,24 +64,37 @@ class SetupInterface
 
   /**
-   * Returns the available options for the 'execute_on' input parameters
+   * (DEPRECATED) Returns the available options for the 'execute_on' input parameters
```
9 Apps need changes, when those are done I will enable the deprecated warnings and then we can keep them as long as you want.

aeslaughter commented on framework/include/utils/MooseEnumItem.h:
```diff
@@ -80,6 +84,9 @@ class MooseEnumItem
   friend std::ostream & operator<<(std::ostream & out, const MooseEnumItem & item);
 
+  // MooseApp::registerExecFlag is allowed to set the id
+  friend MooseApp;
```
I will come up with something better.

aeslaughter commented on framework/include/utils/MooseEnumItem.h:
```diff
@@ -88,7 +95,7 @@ class MooseEnumItem
 
   /// The numeric value for item
-  int _id;
+  mutable int _id; // MooseApp::registerExecFlag can change this for ExecFlagTypesg
```
The class (Moose::execute_flags) can't update itself, because the s (e.g., EXEC_LINEAR) is a global const that is added to execute_flags. The ID of this global object needs to be incremented as it is inserted into the the execute_flags object.

aeslaughter commented on framework/src/base/Moose.C:
```diff
@@ -479,6 +479,26 @@
+const ExecFlagType EXEC_FORCED("FORCED", 0x40);                 // 64
+const ExecFlagType EXEC_FAILED("FAILED", 0x80);                 // 128
+const ExecFlagType EXEC_CUSTOM("CUSTOM", 0x100);                // 256
+const ExecFlagType EXEC_SUBDOMAIN("SUBDOMAIN", 0x200);          // 512
```
They are used in RattleSnake, I can remove them once it is updated.

aeslaughter commented on framework/include/utils/MultiMooseEnum.h:
```diff
@@ -16,12 +16,15 @@
 #include <vector>
 
 // Forward declarations
+class MultiMooseEnum;
```
```
/Users/slauae/projects/moose/framework/include/utils/RandomInterface.h:89:16: error: field has incomplete type 'ExecFlagType' (aka 'MooseEnumItem')
  ExecFlagType _reset_on;
```

aeslaughter commented on framework/include/utils/RandomInterface.h:
```diff
@@ -16,6 +16,7 @@
 #define RANDOMINTERFACE_H
 
 #include "MooseTypes.h"
+#include "MooseEnumItem.h"
```
```
/Users/slauae/projects/moose/framework/include/utils/RandomInterface.h:76:16: error: incomplete result type 'ExecFlagType' (aka 'MooseEnumItem') in function definition
  ExecFlagType getResetOnTime() const { return _reset_on; }
```

permcody commented on framework/src/base/MooseApp.C:
```diff
@@ -1324,3 +1325,28 @@ MooseApp::createMinimalApp()
+MooseApp::registerExecFlag(const ExecFlagType & flag)
+{
+  if (flag.id() == MooseEnumItem::INVALID_ID)
+    flag._id = Moose::execute_flags.getNextValidID();
```
Ok, in response to the "update itself comment". This is where you are needing the friend class and mutable parameter from what I can see. Isn't `getNextValidID()` inherited in the `ExecFlagType` class? If so, can you not just tell the id to update itself using that method?

aeslaughter replied:
`getNextValidID` is *not* part of ExecFlagType (MooseEnumItem), it is part of the ExecFlagEnum (MultiMooseEnum).

aeslaughter commented on framework/include/utils/MultiMooseEnum.h:
```diff
@@ -163,6 +165,7 @@ class MultiMooseEnum : public MooseEnumBase
   // InputParameters and Output is allowed to create an empty enum but is responsible for
   // filling it in after the fact
   friend class libMesh::Parameters;
+  friend class SetupInterface;
```
This is just left-over, I will remove it.

aeslaughter commented on framework/include/utils/RandomData.h:
```diff
@@ -18,6 +18,7 @@
 // MOOSE includes
 #include "MooseRandom.h"
 #include "MooseTypes.h"
+#include "MooseEnumItem.h"
```
```
/Users/slauae/projects/moose/framework/include/utils/RandomInterface.h:89:16: error: field has incomplete type 'ExecFlagType' (aka 'MooseEnumItem')
  ExecFlagType _reset_on;
```

aeslaughter committed:
Address review comments

(refs #7489)

aeslaughter committed:
Alternate method for auto setting execute flag id

(refs #7489)

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/137523/) on 85aa1e7 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 3a8ac7c2695b0b286f4fa82677f47f5d72cfda56`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10110/style.patch).
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10110/style.patch | git apply -v`

permcody commented on framework/include/utils/InputParameters.h:
```diff
@@ -1303,4 +1318,10 @@ InputParameters::getParamHelper(const std::string & name,
 
 InputParameters emptyInputParameters();
 
+// This is temporary specialization to allow applications to utilize MultiMooseEnum for
```
I can understand why you don't have deprecated warnings in here, but do you have a plan to come find all these temporary workarounds to take them out later? Perhaps a searchable tag in the source? Do you have a ticket open for this task?

aeslaughter replied:
The steps will be:
1. Merge this.
2. Update apps for changes 
3. Enable these deprecated messages 
4. Remove deprecated messages after some time

I will open a ticket when I get to 3.

aeslaughter replied:
I will also add a tag of some sort so I can find them all.

permcody commented on framework/src/utils/FormattedTable.C:
```diff
@@ -247,7 +247,21 @@ FormattedTable::printTable(std::ostream & out,
+    {
+      // convert the name into a number, without this conversion the id is used which
+      // is assigned automatically and is not related to the name
+      term_width = std::stoi(suggested_term_width);
```
I'm pretty sure this isn't good enough. @rwcarlsen - Didn't you run into this same problem very recently? This will still partially convert a number that's part of a string. We've solved this in other places in the framework. Take a look at all the boundary and block conversion logic in MooseMesh for instance.

aeslaughter replied:
You are right, the following results in "num = 200".

```
#include <iostream>
#include <string>
int main ()
{
  std::string foo = "200letters";
  int num = std::stoi(foo);
  std::cout << "num = " << num << std::endl;
  return 0;
}
```

permcody replied:
Robert and I have both independently fixed this. I used stringstream and checked for the right flag, but Robert just did this with stoi.

permcody reviewed (dismissed):
Looks good, just two more minor comments. If possible can you avoid rebasing the whole stack of commits when you push up the last commit? That'll make it really easy for me to go through this one last time.

aeslaughter committed:
Add TODO flags for future removal of deprecated code

(refs #7489)

aeslaughter committed:
Improved handling of string to int conversion in FormattedTable

(refs #7489)

permcody commented on framework/src/utils/FormattedTable.C:
```diff
@@ -248,19 +248,15 @@ FormattedTable::printTable(std::ostream & out,
-    }
-    catch (const std::invalid_argument &)
-    {
+    if (std::string(suggested_term_width).find_first_not_of("0123456789") != std::string::npos)
```
Nope! This incorrectly passes an empty string which will fail in stoi below.

Robust solution here: https://github.com/idaholab/moose/blob/devel/framework/src/mesh/MooseMesh.C#L996-L1009

permcody replied:
Ugh... That's not the cleanest example. What you need to watch for is this condition:
```
// If the `>>` operator on the string stream returns false or if the `eof` bit is not set on the stream, something went wrong.
if (!(ss >> id) || !ss.eof())
```

aeslaughter committed:
Update new objects after rebase
(refs #7489)

permcody commented on framework/src/utils/MooseUtils.C:
```diff
@@ -561,6 +564,46 @@ toUpper(const std::string & name)
+  T output;
+  std::istringstream ss(input);
+  if (!(ss >> output) || !ss.eof())
+    mooseError(
```
Well this isn't going to work for the general case. We routinely allow this conversion to fail and are fine with that. Every time you supply a boundary or block like "left" for instance. If this utility is going to be used, you'll have to throw an exception here because you can't presume that users will always want to fail. Also, we have many different error messages that are more specific than this one.

aeslaughter replied:
What would be a good alternative without doing the conversion twice? Could this return a std::pair<T, bool>, where the bool provides pass/fail result?

permcody replied:
You could return a pair, but what would be the value of T when the conversion fails, undefined? I'd just throw, that's cleaner. You don't have to catch, you can just let the error make MOOSE crash.

permcody commented on framework/src/utils/MooseUtils.C:
```diff
@@ -561,6 +564,46 @@ toUpper(const std::string & name)
+template unsigned long long int stringToNumber<unsigned long long int>(const std::string & input);
+template float stringToNumber<float>(const std::string & input);
+template double stringToNumber<double>(const std::string & input);
+template long double stringToNumber<long double>(const std::string & input);
```
I wouldn't bother with this. This utility isn't used too often. Let's just move the template to the header file and remove these.

aeslaughter replied:
I disagree, I don't like to create a template function that can be called with types that don't work. But, if you don't want them I will pull them out.

permcody replied:
That's not what happens with template functions. You only need explicit instantiation if you put the body in the .C file. If it's in the header file, all necessary instantiations are generated on-demand so it works with every type.

aeslaughter replied:
I agree, but it doesn't work with every type. What if I give is a MooseObject? I know that it is stupid, but you can do it.

permcody replied:
Still works as long as the definition is in the header file.

aeslaughter committed:
Don't get too fancy with template function

(refs #7489)

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/139729/) on 04f6d5a wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 941108242af89c1045eb551af583bd4621c88312`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10110/style.patch).
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10110/style.patch | git apply -v`

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/139819/) on 251314f wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 941108242af89c1045eb551af583bd4621c88312`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10110/style.patch).
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10110/style.patch | git apply -v`

aeslaughter committed:
Create stringToInteger to replace stringToNumber that can throw or error

(refs #7489)

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/139888/) on ea027ca wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 941108242af89c1045eb551af583bd4621c88312`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10110/style.patch).
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10110/style.patch | git apply -v`

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/139910/) on 00a68e9 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 941108242af89c1045eb551af583bd4621c88312`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/10110/style.patch).
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10110/style.patch | git apply -v`

permcody commented on framework/src/utils/MooseUtils.C:
```diff
@@ -578,6 +578,40 @@ getDefaultExecFlagEnum()
 }
+
+int
+stringToInteger(const std::string & input, bool throw_on_failure)
```
Well this works, but it seems you took the hard way to get it done. I'll make a suggestion and see what you think.

permcody commented on framework/src/utils/MooseUtils.C:
```diff
@@ -578,6 +578,40 @@ getDefaultExecFlagEnum()
+  // Attempt to use std::stoi, if it fails throw or produce a mooseError
+  try
+  {
+    output = std::stoi(input, &count);
```
Right after this line, I'd do the check on size versus count (currently at the bottom of this routine). If that check fails, throw an invalid_argument. It's fine to test count here and you know there's no error because you'd have already left the try block if there was a problem. By doing this you can localize count and you can also drop the lambda since you can just print the message or fail directly in the catch block since both types of error cases will be sent to the same location (stoi will throw on failure and your size check will also throw when different). That'll cut this routine by more than 50% and make it easier to follow and understand.

aeslaughter replied:
You are right, not sure why I didn't think of that.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/142168/) on 0b4a56d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10110/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10110/style.patch | git apply -v`

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/142179/) on a395e40 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10110/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10110/style.patch | git apply -v`

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/142201/) on 0fa6691 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10110/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10110/style.patch | git apply -v`

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/142212/) on 369023e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10110/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10110/style.patch | git apply -v`

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/142285/) on fb0f58e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10110/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10110/style.patch | git apply -v`

permcody reviewed (approved):
Finally....

rwcarlsen commented:
FWIW this slowed down our benchmarks by about 4-5%.
